compiler/noirc_frontend/src/hir_def/types.rs

```diff
@@ -1737,47 +1745,78 @@ impl Type {
+    fn try_simplify_non_constants(lhs: &Type, op: BinaryTypeOperator, rhs: &Type) -> Option<Type> {
+        let Type::InfixExpr(l_type, l_op, l_rhs) = lhs.follow_bindings() else {
+            return None;
+        };
+
+        // Note that this is exact, syntactic equality, not unification.
+        if l_op.inverse() != Some(op) || l_rhs.canonicalize() != rhs.canonicalize() {
+            return None;
         }
+
+        Some(*l_type)
+    }
+
+    /// Given:
+    ///   lhs = `N op C1`
+    ///   rhs = C2
+    /// Returns: `(N, op, C1, C2)` if C1 and C2 are constants.
+    ///   Note that the operator here is within the `lhs` term, the operator
+    ///   separating lhs and rhs is not needed.
+    fn parse_constant_expr(
+        lhs: &Type,
+        rhs: &Type,
+    ) -> Option<(Box<Type>, BinaryTypeOperator, u32, u32)> {
+        let rhs = rhs.evaluate_to_u32()?;
+
+        let Type::InfixExpr(l_type, l_op, l_rhs) = lhs.follow_bindings() else {
+            return None;
+        };
+
+        let l_rhs = l_rhs.evaluate_to_u32()?;
+        Some((l_type, l_op, l_rhs, rhs))
     }
 
-    /// Try to simplify a subtraction expression of `lhs - rhs`.
+    /// Try to simplify partially constant expressions in the form `(N op1 C1) op2 C2`
+    /// where C1 and C2 are constants that can be combined (e.g. N + 5 - 3 = N + 2)
     ///
-    /// - Simplifies `(a + C1) - C2` to `a + (C1 - C2)` if C1 and C2 are constants.
-    fn try_simplify_subtraction(lhs: &Type, op: BinaryTypeOperator, rhs: &Type) -> Option<Type> {
+    /// - Simplifies `(N +/- C1) +/- C2` to `N +/- (C1 +/- C2)` if C1 and C2 are constants.
+    /// - Simplifies `(N */÷ C1) */÷ C2` to `N */÷ (C1 */÷ C2)` if C1 and C2 are constants.
+    fn try_simplify_partial_constants(
+        lhs: &Type,
+        mut op: BinaryTypeOperator,
+        rhs: &Type,
+    ) -> Option<Type> {
         use BinaryTypeOperator::*;
-        match lhs {
-            Type::InfixExpr(l_lhs, l_op, l_rhs) => {
-                // Simplify `(N + 2) - 1`
-                if op == Subtraction && *l_op == Addition {
-                    if let (Some(lhs_const), Some(rhs_const)) =
-                        (l_rhs.evaluate_to_u32(), rhs.evaluate_to_u32())
-                    {
-                        if lhs_const > rhs_const {
-                            let constant = Box::new(Type::Constant(lhs_const - rhs_const));
-                            return Some(
-                                Type::InfixExpr(l_lhs.clone(), *l_op, constant).canonicalize(),
-                            );
-                        }
-                    }
+        let (l_type, l_op, l_const, r_const) = Type::parse_constant_expr(lhs, rhs)?;
+
+        match (l_op, op) {
+            (Addition | Subtraction, Addition | Subtraction) => {
+                // If l_op is a subtraction we want to inverse the rhs operator.
+                if l_op == Subtraction {
+                    op = op.inverse()?;
```

Comment:
```suggestion
                    op = Addition;
```
`op` will always be addition

Reply:
`op` can also be a subtraction. We can reduce `N - C1 - C2` to `N - (C1 + C2)`